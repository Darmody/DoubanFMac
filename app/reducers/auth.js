import Immutable from 'seamless-immutable';
import { CALL_API } from 'redux-api-middleware';
import { handleActions } from 'redux-actions';
import config from '../../config';

export const LOGIN_REQUEST = 'AUTH/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'AUTH/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'AUTH/LOGIN_FAIL';
export const LOGOUT = 'AUTH/LOGOUT';

const initialState = Immutable({
  user: {
    id: 0
  }
});

export default handleActions({
  [LOGOUT]: (state) => {
    try {
      const PersistStorage = require('electron-json-storage');
      PersistStorage.remove(config.electronStorageKey);
    } catch (error) {
      console.log('Electron Storage Error:', error);
    }

    return {
      ...state,
      user: {
        ...state.user,
        id: 0
      }
    };
  },
  [LOGIN_REQUEST]: (state) => ({
    ...state,
    logged: false,
  }),
  [LOGIN_SUCCESS]: (state, action) => {
    const data = action.payload;

    if (data !== null) {
      return {
        ...state,
        user: {
          id: data.uid,
          name: data.name,
        },
        logged: true,
      };
    }

    return {
      ...state,
      logged: false
    };
  }
}, initialState);

export function logout() {
  return {
    type: LOGOUT
  };
}

export function login(data) {
  const defaultParams = {
    source: 'radio',
    task: 'sync_channel_list',
  };

  const loginFailed = (error) => {
    return {
      type: LOGIN_FAILURE,
      error
    };
  };


  return dispatch => {
    dispatch({
      [CALL_API]: {
        endpoint: 'http://douban.fm/j/login',
        method: 'POST',
        body: {
          ...defaultParams,
          ...data
        },
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        credentials: 'include',
        types: [
          LOGIN_REQUEST,
          {
            type: LOGIN_SUCCESS,
            payload: (action, state, response) => {
              return response.json().then(json => {
                if (json.err_msg) {
                  dispatch(loginFailed(json.err_msg));
                  return null;
                }
                return json.user_info;
              });
            },
          },
          LOGIN_FAILURE
        ]
      }
    });
  };
}
