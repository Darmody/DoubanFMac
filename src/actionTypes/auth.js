export const VERIFY_REQUEST = 'AUTH/VERIFY_REQUEST';
export const VERIFY_SUCCESS = 'AUTH/VERIFY_SUCCESS';
export const VERIFY_FAILURE = 'AUTH/VERIFY_FAILURE';
export const LOGIN_REQUEST = 'AUTH/LOGIN_REQUEST';
export const LOGIN_SUCCESS = 'AUTH/LOGIN_SUCCESS';
export const LOGIN_FAILURE = 'AUTH/LOGIN_FAIL';
export const LOGOUT = 'AUTH/LOGOUT';
