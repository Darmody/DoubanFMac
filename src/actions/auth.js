// @flow
import * as types from 'constants/types/ActionTypes'
import type { Action } from 'constants/types/Redux'

export const login = (username: string, password: string): Action => ({
  type: types.LOGIN_REQUEST,
  payload: {
    username,
    password,
  }
})

export const logout = (): Action => ({
  type: types.LOGOUT,
  payload: undefined,
})

export const logined = (response: Object): Action => ({
  type: types.LOGIN_SUCCESS,
  payload: response,
})
