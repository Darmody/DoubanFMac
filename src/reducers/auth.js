// @flow
import Immutable from 'seamless-immutable'
import * as types from 'constants/types/ActionTypes'
import type { Reducer } from 'constants/types/Redux'

const initialState = Immutable({
  id: null,
  token: '',
  refreshToken: '',
  expiredAt: null,
})

const reducer: Reducer = (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOGINED: return state.merge({
      token: payload.access_token,
      refreshToken: payload.refresh_token,
      id: payload.douban_user_id,
      expiredAt: payload.expires_in,
    })
    default: return state
  }
}

export default reducer
