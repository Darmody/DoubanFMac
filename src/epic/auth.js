// @flow
import { Observable } from 'rxjs'
import * as types from 'constants/types/ActionTypes'
import type { Epic } from 'constants/types/Redux'
import { authorize } from 'clients/doubanRxClient'
import { logined } from 'actions/auth'
import { current } from 'actions/users'

const login = ({ payload }) => authorize(
  payload.username || payload.refreshToken, payload.password
)

const dispatchPersistAndFetch = response => Observable.merge(
  Observable.of(logined(response)),
  Observable.of(current(response.access_token)),
)

const loginEpic: Epic = action$ => action$
  .ofType(types.LOGIN_REQUEST)
  .mergeMap(login)
  .pluck('response')
  .mergeMap(dispatchPersistAndFetch)

export default [
  loginEpic
]
