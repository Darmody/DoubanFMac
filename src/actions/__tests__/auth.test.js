import * as actions from '../auth'

describe('action::auth', () => {
  it('创建一个登录 action', () => {
    expect(actions.login('username', 'password')).toMatchSnapshot()
  })

  it('创建一个登录成功的 action', () => {
    expect(actions.logined({ access_token: 'token' })).toMatchSnapshot()
  })

  it('创建一个登出 action', () => {
    expect(actions.logout()).toMatchSnapshot()
  })

  it('创建一个认证 action', () => {
    expect(actions.auth()).toMatchSnapshot()
  })
})
