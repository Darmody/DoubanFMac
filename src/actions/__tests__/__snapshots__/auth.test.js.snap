// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action::auth 创建一个登录 action 1`] = `
Object {
  "payload": Object {
    "password": "password",
    "username": "username",
  },
  "type": "LOGIN/REQUEST",
}
`;

exports[`action::auth 创建一个登录成功的 action 1`] = `
Object {
  "payload": Object {
    "access_token": "token",
  },
  "type": "LOGIN/SUCCESS",
}
`;
